test/cmocka/src/math/fft/fft.c

```diff
@@ -0,0 +1,406 @@
+	struct icomplex32 *out = (struct icomplex32 *)sink->stream.addr;
+
+	(void)state;
+
+	for (i = 0; i < fft_size; i++)
+		*((int32_t *)source->stream.addr + i) = input_samples[i];
+
+	source->stream.channels = 1;
+	/* do fft transform */
+	fft_real(source, sink, fft_size);
+
+	/* looking for the peak */
+	for (i = 0; i < fft_size / 2; i++) {
+		if ((int64_t)out[i].real * out[i].real +
+		    (int64_t)out[i].imag * out[i].imag > peak) {
+			peak = (int64_t)out[i].real * out[i].real +
+			       (int64_t)out[i].imag * out[i].imag;
+			r = i;
+		}
+	}
+
+	i = (SINE_HZ * fft_size) / 48000;
+	printf("%s: peak at point %d\n", __func__, r);
+
+	/* the peak should be in range i +/-1 */
+	assert_in_range(r, i - 1, i + 1);
+}
+
+static void test_math_fft_1024(void **state)
+{
+	uint32_t fft_size = 1024;
+	int i;
+	int r = 0;
+	int64_t peak = 0;
+	struct sof_ipc_buffer test_buf_desc = {
+		.size = 1024 * 4 * 2,
+	};
+	struct comp_buffer *source = buffer_new(&test_buf_desc);
+	struct comp_buffer *sink = buffer_new(&test_buf_desc);
+	struct icomplex32 *out = (struct icomplex32 *)sink->stream.addr;
+
+	(void)state;
+
+	for (i = 0; i < fft_size; i++)
+		*((int32_t *)source->stream.addr + i) = input_samples[i];
+
+	source->stream.channels = 1;
+	/* do fft transform */
+	fft_real(source, sink, fft_size);
+
+	/* looking for the peak */
+	for (i = 0; i < fft_size / 2; i++) {
+		if ((int64_t)out[i].real * out[i].real +
+		    (int64_t)out[i].imag * out[i].imag > peak) {
+			peak = (int64_t)out[i].real * out[i].real +
+			       (int64_t)out[i].imag * out[i].imag;
+			r = i;
+		}
+	}
+
+	i = (SINE_HZ * fft_size) / 48000;
+	printf("%s: peak at point %d\n", __func__, r);
+
+	/* the peak should be in range i +/-1 */
+	assert_in_range(r, i - 1, i + 1);
+}
+
+static void test_math_fft_1024_ifft(void **state)
+{
+	uint32_t fft_size = 1024;
+	int i;
+	int64_t signal = 0;
+	int64_t noise = 0;
+	int r = 0;
+	float db;
+	struct sof_ipc_buffer test_buf_desc = {
+		.size = 1024 * 4 * 2,
+	};
+	struct comp_buffer *source = buffer_new(&test_buf_desc);
+	struct comp_buffer *intm = buffer_new(&test_buf_desc);
+	struct comp_buffer *sink = buffer_new(&test_buf_desc);
+	struct icomplex32 *out = (struct icomplex32 *)sink->stream.addr;
+
+	(void)state;
+
+	for (i = 0; i < fft_size; i++)
+		*((int32_t *)source->stream.addr + i) = input_samples[i];
+
+	source->stream.channels = 1;
+	/* do fft transform */
+	fft_real(source, intm, fft_size);
+
+	intm->stream.channels = 1;
+	/* do ifft transform */
+	ifft_complex(intm, sink, fft_size);
+
+	/* calculate signal and noise */
+	for (i = 0; i < fft_size; i++) {
+		signal += (int64_t)(input_samples[i] / 32) * (input_samples[i] / 32);
+		noise += (int64_t)((out[i].real - input_samples[i]) / 32) *
+			 ((out[i].real - input_samples[i]) / 32) +
+			 (int64_t)(out[i].imag / 32) * (out[i].imag / 32);
+	}
+
+	/* db in Q8.24 */
+	db = 10 * log10((float)signal / noise);
+
+	printf("%s: signal: 0x%llx noise: 0x%llx db: %f\n", __func__,
```

Comment:
Would be nice to compare total harmonic distortion and signal to noise and distortion ration

Reply:
> Would be nice to compare total harmonic distortion and signal to noise and distortion ration

since we are comparing time domain samples here, it is not easy to get the frequency domain THD?